File: staging/ABF-001/mathematical-source/src/abf001_small_universe.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define FUNCTIONS (1u << 16)
#define EXPECTED_COMPARISONS (65536ull * 14ull * 3ull * 2ull)

static inline unsigned parity_u32(uint32_t x) {
    return (unsigned)(__builtin_popcount(x) & 1u);
}

static void kernel_basis(unsigned a, unsigned *u_out, unsigned *v_out) {
    unsigned u = 0, v = 0;
    for (unsigned x = 1; x < 8; ++x) {
        if (parity_u32(a & x) == 0u) {
            if (u == 0u) u = x;
            else if (x != u) { v = x; break; }
        }
    }
    if (u == 0u || v == 0u || u == v) {
        fprintf(stderr, "kernel basis failure for a=%u\n", a);
        exit(2);
    }
    *u_out = u;
    *v_out = v;
}

static void affine_points(unsigned a, unsigned b, unsigned points[4]) {
    unsigned p0 = 8;
    for (unsigned x = 0; x < 8; ++x) {
        if (parity_u32(a & x) == b) { p0 = x; break; }
    }
    if (p0 == 8) { fprintf(stderr, "no affine point\n"); exit(2); }
    unsigned u, v;
    kernel_basis(a, &u, &v);
    points[0] = p0;
    points[1] = p0 ^ u;
    points[2] = p0 ^ v;
    points[3] = p0 ^ u ^ v;
    for (unsigned i = 0; i < 4; ++i) {
        for (unsigned j = i + 1; j < 4; ++j) {
            if (points[i] == points[j]) {
                fprintf(stderr, "duplicate point\n"); exit(2);
            }
        }
    }
}

static inline unsigned scalar_value(uint16_t function, unsigned x, unsigned mask) {
    unsigned output = (function >> (2u * x)) & 3u;
    return parity_u32(output & mask);
}

static int anf_degree_four(const unsigned values_in[4]) {
    unsigned c0 = values_in[0];
    unsigned c1 = values_in[1] ^ values_in[0];
    unsigned c2 = values_in[2] ^ values_in[0];
    unsigned c3 = values_in[3] ^ values_in[2] ^ values_in[1] ^ values_in[0];
    int degree = -1;
    if (c0) degree = 0;
    if (c1 || c2) degree = 1;
    if (c3) degree = 2;
    return degree;
}

static int moments_vanish(uint16_t function, const unsigned points[4], unsigned mask, unsigned order) {
    unsigned total = 0;
    for (unsigned i = 0; i < 4; ++i) total ^= scalar_value(function, points[i], mask);
    if (total) return 0;
    if (order == 0u) return 1;
    for (unsigned coordinate = 0; coordinate < 3; ++coordinate) {
        total = 0;
        for (unsigned i = 0; i < 4; ++i) {
            if ((points[i] >> coordinate) & 1u) total ^= scalar_value(function, points[i], mask);
        }
        if (total) return 0;
    }
    return 1;
}

int main(void) {
    unsigned hpoints[14][4];
    unsigned h = 0;
    for (unsigned a = 1; a < 8; ++a) {
        for (unsigned b = 0; b < 2; ++b) {
            affine_points(a, b, hpoints[h]);
            ++h;
        }
    }
    uint64_t comparisons = 0;
    uint64_t failures = 0;
    clock_t start = clock();
    for (uint32_t function = 0; function < FUNCTIONS; ++function) {
        for (unsigned hi = 0; hi < 14; ++hi) {
            const unsigned *points = hpoints[hi];
            for (unsigned mask = 1; mask <= 3; ++mask) {
                unsigned values[4];
                for (unsigned i = 0; i < 4; ++i) values[i] = scalar_value((uint16_t)function, points[i], mask);
                int degree = anf_degree_four(values);
                for (unsigned order = 0; order <= 1; ++order) {
                    int threshold = 1 - (int)order;
                    int direct = degree <= threshold;
                    int moment = moments_vanish((uint16_t)function, points, mask, order);
                    ++comparisons;
                    if (direct != moment) {
                        ++failures;
                        if (failures <= 20) {
                            fprintf(stderr, "mismatch function=%04x hyperplane=%u mask=%u order=%u degree=%d direct=%d moment=%d\n",
                                    function, hi, mask, order, degree, direct, moment);
                        }
                    }
                }
            }
        }
    }
    double elapsed = (double)(clock() - start) / (double)CLOCKS_PER_SEC;
    if (comparisons != EXPECTED_COMPARISONS || failures != 0) {
        fprintf(stderr, "FAIL comparisons=%llu expected=%llu failures=%llu\n",
                (unsigned long long)comparisons,
                (unsigned long long)EXPECTED_COMPARISONS,
                (unsigned long long)failures);
        return 1;
    }
    printf("{\n");
    printf("  \"schema_version\": \"n.human_llm.mathematics.abf001.small_universe_c.v1\",\n");
    printf("  \"result\": \"PASS\",\n");
    printf("  \"maps\": 65536,\n");
    printf("  \"affine_hyperplanes_per_map\": 14,\n");
    printf("  \"nonzero_output_masks\": 3,\n");
    printf("  \"moment_orders\": [0, 1],\n");
    printf("  \"comparisons\": %llu,\n", (unsigned long long)comparisons);
    printf("  \"failures\": 0,\n");
    printf("  \"elapsed_cpu_seconds\": %.6f\n", elapsed);
    printf("}\n");
    return 0;
}

File: staging/ABF-001/mathematical-source/src/abf001_small_universe.py
#!/usr/bin/env python3
"""Complete small-universe control for the ABF-001 moment criterion.

Enumerates all 65,536 vectorial maps F: GF(2)^3 -> GF(2)^2, all 14 affine
hyperplanes, all three nonzero output masks, and moment orders r=0,1.
It compares direct restriction ANF degree with ambient moment vanishing.
Total exact comparisons: 65,536 * 14 * 3 * 2 = 5,505,024.

This implementation is standalone and imports neither ABF primary verifier.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import platform
import sys
import time
from pathlib import Path

FUNCTIONS = 1 << 16  # two output bits at each of eight input points
NORMALS = tuple(range(1, 8))
OFFSETS = (0, 1)
MASKS = (1, 2, 3)
ORDERS = (0, 1)
EXPECTED_COMPARISONS = FUNCTIONS * len(NORMALS) * len(OFFSETS) * len(MASKS) * len(ORDERS)


def parity(x: int) -> int:
    return x.bit_count() & 1


def kernel_basis(a: int) -> tuple[int, int]:
    vectors = [x for x in range(1, 8) if parity(a & x) == 0]
    u = vectors[0]
    v = next(x for x in vectors[1:] if x != u)
    if v == u:
        raise AssertionError
    return u, v


def affine_points(a: int, b: int) -> tuple[int, int, int, int]:
    p0 = next(x for x in range(8) if parity(a & x) == b)
    u, v = kernel_basis(a)
    points = (p0, p0 ^ u, p0 ^ v, p0 ^ u ^ v)
    if len(set(points)) != 4:
        raise AssertionError((a, b, points))
    return points


def scalar_value(function: int, x: int, mask: int) -> int:
    output = (function >> (2 * x)) & 3
    return parity(output & mask)


def anf_degree_four(values: tuple[int, int, int, int]) -> int:
    coeff = list(values)
    coeff[1] ^= coeff[0]
    coeff[3] ^= coeff[2]
    coeff[2] ^= coeff[0]
    coeff[3] ^= coeff[1]
    degrees = (-1, 1, 1, 2)
    degree = -1
    for index, value in enumerate(coeff):
        if value:
            degree = max(degree, degrees[index] if index else 0)
    return degree


def moments_vanish(function: int, a: int, b: int, mask: int, order: int) -> bool:
    points = affine_points(a, b)
    # empty monomial
    total = 0
    for x in points:
        total ^= scalar_value(function, x, mask)
    if total:
        return False
    if order == 0:
        return True
    for coordinate in range(3):
        total = 0
        for x in points:
            if (x >> coordinate) & 1:
                total ^= scalar_value(function, x, mask)
        if total:
            return False
    return True


def run() -> dict:
    start = time.perf_counter()
    comparisons = 0
    failures: list[dict] = []
    hyperplanes = [(a, b, affine_points(a, b)) for a in NORMALS for b in OFFSETS]

    for function in range(FUNCTIONS):
        for a, b, points in hyperplanes:
            for mask in MASKS:
                values = tuple(scalar_value(function, x, mask) for x in points)
                degree = anf_degree_four(values)
                for order in ORDERS:
                    threshold = 1 - order
                    direct = degree <= threshold
                    moment = moments_vanish(function, a, b, mask, order)
                    comparisons += 1
                    if direct != moment:
                        failures.append({
                            "function_hex": f"{function:04x}",
                            "normal": a,
                            "offset": b,
                            "mask": mask,
                            "order": order,
                            "degree": degree,
                            "direct": direct,
                            "moment": moment,
                        })
                        if len(failures) >= 20:
                            raise AssertionError(failures)

    elapsed = time.perf_counter() - start
    if comparisons != EXPECTED_COMPARISONS:
        raise AssertionError((comparisons, EXPECTED_COMPARISONS))
    if failures:
        raise AssertionError(failures)

    report = {
        "schema_version": "n.human_llm_mathematics.abf001.small_universe.v1",
        "result": "PASS",
        "universe": {
            "maps": FUNCTIONS,
            "affine_hyperplanes_per_map": 14,
            "nonzero_output_masks": 3,
            "moment_orders": [0, 1],
            "comparisons": comparisons,
        },
        "failures": 0,
        "implementation": "standalone direct restriction ANF versus ambient moment sums",
        "python": sys.version,
        "platform": platform.platform(),
        "elapsed_seconds": round(elapsed, 6),
    }
    canonical = json.dumps(report, sort_keys=True, separators=(",", ":")).encode("utf-8")
    report["evidence_sha256"] = hashlib.sha256(canonical).hexdigest()
    return report


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    report = run()
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(json.dumps(report, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    print(json.dumps(report, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
